### ISSUE TITLE
Shortened constituents Result

### ISSUE THREAD

**Issue by sbguthery** [opencog/link-grammar#991]
The below sentence yields a shortened constituents string when the constituents value is set to 3.  The other results---graphics, disjuncts, etc.--- are fine.  The shortened constituents string is:

(ADVP (S (NP this.d result.n) (VP was.v-d apparently confirmed.v-d (PP by (NP some independent.n experiments.n)))))

The sentence is:

"This result was apparently confirmed by  some independent experiments  but it is very far from the truth  for  it is now known that the actual ratio  or factor as it is commonly  called  of the velocity of the wind to that of the cups depends very  largely on the dimensions of the cups and arms  and may have almost  any value between two and a little over three."

This is for version 5.6.2.

Cheers, Scott

**Comment by ampli** [comment]
I couldn't reproduce that using `link-parser` or python (also using 5.6.2): 
```
$ link-parser
link-grammar: Info: Dictionary found at ../en/4.0.dict
link-grammar: Info: Dictionary version 5.6.2, locale en_US.UTF-8
link-grammar: Info: Library version link-grammar-5.6.2. Enter "!help" for help.
linkparser> !graph
Graphical display of linkage turned off.
linkparser> !const=3
constituents set to 3
linkparser> This result was apparently confirmed by some independent experiments but it is very far from the truth for it is now known that the actual ratio or factor as it is commonly called of the velocity of the wind to that of the cups depends very largely on the dimensions of the cups and arms and may have almost any value between two and a little over three.
No complete linkages found.
link-grammar: Warning: Combinatorial explosion! nulls=1 cnt=2147483647
Consider retrying the parse with the max allowed disjunct cost set lower.
At the command line, use !cost-max
Found 2147483647 linkages (118 of 1000 random linkages had no P.P. violations) at null count 1
	Linkage 1, cost vector = (UNUSED=1 DIS=17.28 LEN=248)
(S (ADVP (S (NP this.d result.n) (VP was.v-d apparently confirmed.v-d (PP by (NP some independent.n experiments.n))))) but.ij (S (NP it) (VP is.v (ADVP very.e far.e) (PP from (NP the truth.s)) for.r (S (NP it) (VP is.v (VP (ADVP now.r) known.v (SBAR that.j-c (S (NP (NP (NP (NP (NP (NP the actual.a ratio.n or.j-n factor.n) (PP as.e (NP it)) {is}) (VP (ADVP commonly) called.v-d)) (PP of (NP the velocity.s)) of the wind.n-u) (PP to.r (NP that.j-p))) (PP of (NP the cups.n))) (VP depends.v very.e largely.e (PP on (NP the dimensions.n of the cups.n and.j-n arms.n)) and.j-v may.v (VP have.v (NP (PP (NP (NP almost any value.n-u) (PP between (NP two)) and.j-n a little) (PP over (NP three))))))))))))) .)
```
```
$ python3
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from linkgrammar import *
>>> print(Sentence("This result was apparently confirmed by some independent experiments but it is very far from the truth for it is now known that the actual ratio or factor as it is commonly called of the velocity of the wind to that of the cups depends very largely on the dimensions of the cups and arms and may have almost any value between two and a little over three.",Dictionary(),ParseOptions(min_null_count=0, max_null_count=999)).parse().next().constituent_tree(3))
(S (S (NP this.d result.n) (VP was.v-d (VP (ADVP apparently) confirmed.v-d (PP by (NP some independent.n experiments.n))))) but.ij (S (NP it) (VP is.v (ADVP very.e far.e) (PP from (NP the truth.s)) for.r (S (NP it) (VP is.v (VP (ADVP now.r) known.v (SBAR (NP (NP (NP that.j-c (S (NP (NP the actual.a ratio.n or.j-n factor.n) (PP as.e (NP it))) (VP is.v (VP (ADVP commonly) called.v-d))) {of} the velocity.s of the wind.n-u) (PP to.r (NP that.j-p))) (PP of (NP the cups.n))) (VP depends.v very.e largely.e (PP on the dimensions.n (PP of (NP the cups.n)) and.j-n arms.n) and.j-v may.v (VP have.v (NP (NP almost any value.s) (PP between (NP two)) and.j-n a little) (PP over (NP three)))))))))) .)
```
(The different constituent tree in these 2 tests is because the default parse-options of the link-grammar Python bindings are different than of `link-parser`.)

So I guess you used something else, or maybe different parse-options.
Please post the code you used.

**Comment by sbguthery** [comment]
Hello, Amir ...

My apologies for not documenting the experimental conditions.

The attached input.txt when given to link-parser

./link-parser < input.txt

generates the attached output.txt.

Cheers, Scott

link-grammar: Info: Dictionary found at ../data/en/4.0.dict
link-grammar: Info: Dictionary version 5.6.2, locale en_US.UTF-8
link-grammar: Info: Library version link-grammar-5.6.2. Enter "!help" for help.
verbosity set to 0
Use of null-linked islands turned on.
graphics set to 1
Echoing of input sentence turned on.
limit set to 1
Display of disjuncts used turned on.
constituents set to 3
This result was apparently confirmed by  some independent experiments  but it is very far from the truth  for  it is now known that the actual ratio  or factor as it is commonly  called  of the velocity of the wind to that of the cups depends very  largely on the dimensions of the cups and arms  and may have almost  any value between two and a little over three.


    +----------------------------------------------Xx--------------
    +---------->WV---------->+                              +------
    +------>Wd-------+       +--------------MVp-------------+   +--
    |        +-Dsu*x-+--Ss*t-+---EBm---+----EAy----+        |   |
    |        |       |       |         |           |        |   |
LEFT-WALL this.d result.n was.v-d apparently confirmed.v-d by some


                               +-------------------------------->WV--------
-------------------------------+        +-----------------VC---------------
------Jp------------+          |        +-------MVp-------+
--------Dmc---------+          |        +----MVa----+     +----Ju----+
      +------AN-----+          +-Wdc+-Ss+     +--EE-+     |   +--Dmu-+
      |             |          |    |   |     |     |     |   |      |
independent.n experiments.n but.ij it is.v very.e far.e from the truth.n-u


----------------------->+
--+                     |        +---------------------------------CV------
  |                     |        +-------------Cet------------+----------Ss
  |         +----Pvf----+        |     +-----Ds**x----+       +------Mp----
  +-Wd-+SFsi+    +---E--+---THi--+     |      +---A---+--SJls-+--SJrs-+
  |    |    |    |      |        |     |      |       |       |       |
for.r it  is.v now.r known.v that.j-c the actual.a ratio.n or.j-n factor.n



-------------------------->+
x---------+                +-------------------MVp-------------------+
--+       +-------Pv-------+---------Os--------+      +---Js---+     |
  +-J-+   +--MVa-+         |           +-Ds**c-+--Mp--+  +Ds**c+--Mp-+--
  |   |   |      |         |           |       |      |  |     |     |
as.e it is.v commonly called.v-d [of] the velocity.s of the wind.s to.r


          +---------------------------------------Js-------------------------
          |          +--------------------------------VJlgi------------------
          |          +------------------------------Opn----------------------
          |          |                               +-----Jp----+
Js--+--Mf-+          +-------MVl------+--MVl--+--MVp-+  +---Dmc--+
    |     |          |                |       |      |  |        |
that.j-p of [the] cups.v [depends] very.e largely.e on the dimensions.n [of]


--------------+
--------------+
-------+      |                +--------------------------Op-------------------
       |      +------VJrg------+                                 +--------SJlp-
 +-Dmc-+      |       +--VJlgt-+--VJrg-+--If--+--MVp-+-Ju-+      +---Mp---+--Jp
 |     |      |       |        |       |      |      |    |      |        |
the cups.n and.j-v arms.v and.v-fill may.v have.v almost any value.n-u between




-------+-----------Xc-----------+
-------+---SJrs--+              |
-+     |    +IDPO+--Mp-+-Jp-+   |
 |     |    |    |     |    |   |
two and.j-n a little over three .

(ADVP (S (NP this.d result.n) (VP was.v-d apparently confirmed.v-d (PP by (NP some independent.n experiments.n)))))

            LEFT-WALL    0.060  hWd+ hWV+ Xx+
               this.d    0.000  D*u+
             result.n    0.000  Ds**x- Wd- Ss*t+
              was.v-d    0.666  Ss- dWV- @EBm+ @MV+
           apparently    -0.100  EB- EA+
        confirmed.v-d    0.000  EAy-
                   by    0.000  MVp- J+
                 some    0.000  D+
        independent.n    0.000  AN+
        experiments.n    0.100  @AN- Dmc- Jp-
               but.ij    1.100  Xx- Wdc+ hWV+
                   it    0.000  Wd- Ss+
                 is.v    1.666  Ss- @MV+ VC+
               very.e    0.000  EE+
                far.e    0.000  EE- MVa-
                 from    0.000  MVp- J+
                  the    0.000  D+
            truth.n-u    0.000  Dmu- Ju-
                for.r    2.000  VC- Wd+
                   it    0.000  Wd- SFsi+
                 is.v    0.100  SFs- Pv+
                now.r    0.000  E+
              known.v    0.000  @E- Pvf- dWV- THi+
             that.j-c    0.000  TH- Cet+ CV+
                  the    0.000  D+
             actual.a    0.000  A+
              ratio.n    0.000  @A- Ds**x- SJls+
               or.j-n    0.000  SJls- C- SJrs+ @M+ S*x+
             factor.n    0.000  SJrs-
                 as.e    1.000  Mp- J+
                   it    1.000  J-
                 is.v    0.100  Ss- MV+ Pv+
             commonly    0.000  MVa-
           called.v-d    0.000  Pv- dCV- O+ @MV+
                  the    0.000  D+
           velocity.s    0.000  Ds**c- Os- @M+
                   of    0.000  Mp- Js+
                  the    0.000  D+
               wind.s    0.000  Ds**c- Js- @M+
                 to.r    0.390  Mp- MVp- J+
             that.j-p    4.000  Js- M+
                   of    0.000  Mf- Js+
               cups.v    2.000  @MV+ O*n+ VJl*i+
               very.e    0.000  MVl- MVl+
            largely.e    0.500  MVl- MVp+
                   on    0.000  MVp- J+
                  the    0.000  D+
         dimensions.n    0.000  Dmc- Jp-
                  the    0.000  D+
               cups.n    1.000  Dmc- Op-
              and.j-v    0.000  VJlg- J- VJrg+
               arms.v    0.000  VJl*t+
           and.v-fill    0.000  VJlg- VJrg- VJrg+ O+
                may.v    0.000  VJr- I+
               have.v    2.000  If- @MV+
               almost    0.000  MVp- Ju+
                  any    0.000  J-
            value.n-u    0.000  Mp+ SJlp+
              between    0.400  Mp- J+
                  two    1.000  Jp-
              and.j-n    3.000  SJlp- Op- SJr+ Xc+
                    a    0.000  IDPO+
               little    1.000  IDPO- SJrs- @M+
                 over    0.400  Mp- J+
                three    1.000  Jp-
                    .    2.000  Xc- RW+
           RIGHT-WALL    0.000  RW-



    +----------------------------------------------Xx--------------
    +---------->WV---------->+                              +------
    +------>Wd-------+       +--------------MVp-------------+   +--
    |        +-Dsu*c-+--Ss*t-+---EBm---+----EAy----+        |   |
    |        |       |       |         |           |        |   |
LEFT-WALL this.d result.n was.v-d apparently confirmed.v-d by some


                               +-------------------------------->WV---------
-------------------------------+         +-----------------VC---------------
------Jp------------+          |         +-------MVp-------+
--------Dmc---------+          |         +----MVa----+     +----Ju----+
      +------AN-----+          +-Wdc+SFsi+     +--EE-+     |   +--Dmu-+
      |             |          |    |    |     |     |     |   |      |
independent.n experiments.n but.ij it  is.v very.e far.e from the truth.n-u


----------------------->+
--+                     |        +-------------------------CV--------------
  |                     |        +-------------Cet------------+
  |         +----Pvf----+        |     +-----Ds**x----+       +----------Ss
  +-Wd-+SFsi+    +---E--+---THi--+     |      +---A---+--SJls-+--SJrs-+--Mp
  |    |    |    |      |        |     |      |       |       |       |
for.r it  is.v now.r known.v that.j-c the actual.a ratio.n or.j-n factor.n



--------->+
          |
x---------+---------OFj---------+-----Jp----+       +----Ju---+---------
--+-J-+   +--EBm-+              |  +---D*u--+---Mp--+  +--Dmu-+--Mp--+--
  |   |   |      |              |  |        |       |  |      |      |
as.e it is.v commonly [called] of the velocity.n-u of the wind.n-u to.r


          +---------------------------------------Jp-------------------------
          |          +--------------------------------VJlgi------------------
          |          +------------------------------Opn----------------------
Mf--------+          |                               +-----Jp----+
Js--+     |          +-------MVl------+--MVl--+--MVp-+  +---Dmc--+
    |     |          |                |       |      |  |        |
that.j-p of [the] cups.v [depends] very.e largely.e on the dimensions.n [of]


--------------+
--------------+
-------+      |                +--------------------------Oun------------------
       |      +------VJrg------+---------MVp---------+           +--------SJlu-
 +-Dmc-+      |       +--VJlgt-+--VJrg-+--If--+      +-Ju-+      +---Mp---+--Jp
 |     |      |       |        |       |      |      |    |      |        |
the cups.n and.j-v arms.v and.v-fill may.v have.v almost any value.n-u between




-------+-----------Xc-----------+
-------+---SJrs--+              |
-+     |    +IDPO+--Mp-+-Jp-+   |
 |     |    |    |     |    |   |
two and.j-n a little over three .

(ADVP (S (NP this.d result.n) (VP was.v-d apparently confirmed.v-d (PP by (NP some independent.n experiments.n)))))

            LEFT-WALL    0.060  hWd+ hWV+ Xx+
               this.d    0.000  D*u+
             result.n    0.000  Ds**c- Wd- Ss*t+
              was.v-d    0.666  Ss- dWV- @EBm+ @MV+
           apparently    -0.100  EB- EA+
        confirmed.v-d    0.000  EAy-
                   by    0.000  MVp- J+
                 some    0.000  D+
        independent.n    0.000  AN+
        experiments.n    0.100  @AN- Dmc- Jp-
               but.ij    1.100  Xx- Wdc+ hWV+
                   it    0.000  Wd- SFsi+
                 is.v    1.666  SFs- @MV+ VC+
               very.e    0.000  EE+
                far.e    0.000  EE- MVa-
                 from    0.000  MVp- J+
                  the    0.000  D+
            truth.n-u    0.000  Dmu- Ju-
                for.r    2.000  VC- Wd+
                   it    0.000  Wd- SFsi+
                 is.v    0.100  SFs- Pv+
                now.r    0.000  E+
              known.v    0.000  @E- Pvf- dWV- THi+
             that.j-c    0.000  TH- Cet+ CV+
                  the    0.000  D+
             actual.a    0.000  A+
              ratio.n    0.000  @A- Ds**x- SJls+
               or.j-n    0.000  SJls- C- SJrs+ S*x+
             factor.n    1.000  SJrs- @M+
                 as.e    1.000  Mp- J+
                   it    1.000  J-
                 is.v    0.000  Ss- dCV- @EBm+ OF+
             commonly    0.000  EBm-
                   of    0.000  OFj- Jp+
                  the    0.000  D+
         velocity.n-u    0.000  D*u- Jp- @M+
                   of    0.000  Mp- Ju+
                  the    0.000  D+
             wind.n-u    0.000  Dmu- Ju- @M+
                 to.r    1.400  Mp- J+
             that.j-p    2.000  Js-
                   of    0.000  Mf- Jp+
               cups.v    2.000  @MV+ O*n+ VJl*i+
               very.e    0.000  MVl- MVl+
            largely.e    0.500  MVl- MVp+
                   on    0.000  MVp- J+
                  the    0.000  D+
         dimensions.n    0.000  Dmc- Jp-
                  the    0.000  D+
               cups.n    1.000  Dmc- Op-
              and.j-v    0.000  VJlg- J- VJrg+
               arms.v    0.000  VJl*t+
           and.v-fill    2.000  VJlg- VJrg- VJrg+ @MV+ O*n+
                may.v    0.000  VJr- I+
               have.v    2.000  If-
               almost    0.000  MVp- Ju+
                  any    0.000  J-
            value.n-u    0.400  @M+ SJlu+
              between    0.400  Mp- J+
                  two    1.000  Jp-
              and.j-n    3.000  SJlu- Ou- SJr+ Xc+
                    a    0.000  IDPO+
               little    1.000  IDPO- SJrs- @M+
                 over    0.400  Mp- J+
                three    1.000  Jp-
                    .    2.000  Xc- RW+
           RIGHT-WALL    0.000  RW-

Bye.

!verbosity=0
!islands-ok
!graphics=1
!echo
!limit=1
!disjuncts
!constituents=3
This result was apparently confirmed by  some independent experiments  but it is very far from the truth  for  it is now known that the actual ratio  or factor as it is commonly  called  of the velocity of the wind to that of the cups depends very  largely on the dimensions of the cups and arms  and may have almost  any value between two and a little over three.

**Comment by linas** [comment]
My quick guess is that constituent processing halted at the end of the first island.  (I'm guessing that ) The problem is that each "island" behaves like a "complete" sentence, and the constituents processing doesn't quite realize there is a "second" (pseudo-) sentence there.  The work-around is to not enable islands. (and instead accept a single sentence with a missing word or two)

**Comment by linas** [comment]
My last remark is kind-of wrong but kind of right. It has nothing to do wit null words or islands -- I tested "This result was apparently confirmed by some independent experiments but it is very far from the truth." which has no  null words or islands, but it is analyzed as being two independent sentences conjoined by the word "but". (The `Xx` points to the start of the "second" sentence.)

I'm unclear as to what the correct phrase structure for something like this should be, anyway ... Should it be `(S This result ... (S but ...))` ?? 

Scott, please be aware that the phrase structure is derived from the link parse ... it is both simpler, but also less flexible, and conveys less information.

**Comment by sbguthery** [comment]
Here's my parsing of the sentence:

This result[SUBJECT]
    1] was [VERB]
        apparently confirmed by some independent experiments [OBJECT/was]

    but it [CONJUNCTION]

    2] is [VERB]
        very far [OBJECT/is]
            from the truth [ADJECTIVE PHRASE/far]
                for it is now known that the actual ratio [ADJECTIVE PHRASE/truth]
                    or factor [ADJECTIVE PHRASE/ratio]
                        as it is commonly called [ADJECTIVE PHRASE/factor]
                of the velocity of the wind to that of the cups [ADJECTIVE PHRASE/ratio]

    3] depends [VERB] very largely 
        on the dimensions of the cups and arms [OBJECT/depends]

    and [CONJUNCTION]

    4] may have [VERB]
        almost  any value [OBJECT/have]
            between two and a little over three [ADJECTIVE PHRASE/value]

In short ...

result [was confirmed] [is far] [depends on] [may have]

Cheers, Scott

P.S. From ANEMOMETER, Encyclopedia Britannica, 11th edition.

**Comment by linas** [comment]
> Here's my parsing of the sentence:
>

There is no standardized, formalized, generally agreed on phrase-structure
grammar rules to which LG can adhere to; the most one can do is to emulate
something like the Pen treebank corpus, which is famously a rather ad hoc
collection replete with arbitrary choices and inconsistencies, created by a
myriad of grad students and academics.  With that in mind, one can try to
proceed using "common sense", which is what you're doing.

>
> This result[SUBJECT]
> 1] was [VERB]
> apparently confirmed by some independent experiments [OBJECT/was]
>
> but it [CONJUNCTION]
>

"but" is the conjunction. "it" is the subject of a new sentence. 100 years
ago, the style was to write very long sentences, by eschewing the period,
and using non-puncutational conjunction words instead.  I read Marx's
"Manuscripts of 1844" in college. There was approximately one period per
each rather long paragraph. Link Grammar can (and mostly does) understand
that the period, just like other punctuation, is just one more kind of
conjunction.

>
> 2] is [VERB]
> very far [OBJECT/is]
> from the truth [ADJECTIVE PHRASE/far]
>

"from" is a prepostion; prepostions tend to work much like verbs.
"the truth" is the prepositional object.
"it is very far" is the prepositional subject.

for it is now known that the actual ratio [ADJECTIVE PHRASE/truth]
> or factor [ADJECTIVE PHRASE/ratio]
>

"for" is a conjunction here; it delimits the end of the previous sentence,
and starts a new one.  The author could have (but didn't) inserted a
sentence-ending period here.

"it is now known that" is a that-phrase.
"ratio or factor" is a conjunction, in the classical sense of two conjoined
nouns.
It is no clear if "the actual" is modifying only the noun "ratio", or if it
is modifying the entire conjoined noun-phrase


> as it is commonly called [ADJECTIVE PHRASE/factor]
>

Yes; this is a post-nominal adjectival phrase. Adjectives normally come
before nouns ("red balloon" not "balloon red") but the "as" delimits the
post-nominal construction.


> of the velocity of the wind to that of the cups [ADJECTIVE PHRASE/ratio]
>

No, because, again, "of" is a preposition.
"ratio or factor" is the prepositional subject,
"the velocity" is the prepositional object of the earlier "of", and the
prepositional subject of the next "of"

"to that of" ... denotes a mathematical formula, viz a relation or
comparison. (comparisons, ratios, divisions, factors, quotients are all
formulaic in this way: "we compare the size of the X to that of the Y"  The
comparison requires two objects (X,Y), together with the dimension being
compared ("the size") and "that" is the anaphora resolution of "the size":
"we compare the size of the X to the size of the Y" is the anaphora.

>
> 3] depends [VERB] very largely
> on the dimensions of the cups and arms [OBJECT/depends]
>
> and [CONJUNCTION]
>

Again, the author could have placed a sentence-ending period here, but
chose not to.

>
> 4] may have [VERB]
>

There is a zero subject here: an unspoken/unwritten "it" that is another
anaphora resolution going back to "ratio or factor".

almost any value [OBJECT/have]
> between two and a little over three [ADJECTIVE PHRASE/value]
>

Again, a post-nominal adjectival phrase!

This is a complex, difficult-to-understand sentence. It is no accident that
increases in literacy in the 20th century are accompanied by a general
shortening of sentences: Edwardian English has fallen out of style;
everyone can read these days, but not everyone can read long, complex
sentences.   What makes this charming is that, if this had been split up
into four distinct sentences, it would be just as hard to understand: there
would still be multiple anaphora resolutions to be performed, except that
they would be between different sentences, rather than between different
clauses in the same sentence.

I like to visualize such sentences as a collection of mechanical gears or
pulleys, with prepositions being like a rope strung between them: you have
to carefully examine what is connected to what, to see how it works, to
*articulate* the linkages in the device.  Syntactic parsers ignore the
articulation between sentences: long sentences force the issue to the
forefront.

**Comment by linas** [comment]
Closing .... Please open a new issue if there are any further problems.
